openedx/core/djangoapps/plugins/plugin_slots.py

```diff
@@ -0,0 +1,114 @@
+from importlib import import_module
+from logging import getLogger
+from typing import Dict, List, Tuple, Callable
+
+from django.apps import AppConfig
+from importlib_metadata import ModuleNotFoundError
+
+from openedx.core.lib.cache_utils import process_cached
+from . import constants, registry
+from .plugin_contexts import get_plugins_view_context
+
+log = getLogger(__name__)
+
+COMMON_ALLOW_LIST = ['request', 'current_url']
+
+
+def get_content_for_slot(project_type: str, slot_namespace: str, slot_name: str, raw_context: Dict) -> str:
+    """
+    Returns a list of additional content for a view slot. Will check if any plugin apps
+    have that view in their view_context_config, and if so will call their
+    selected function to get their context dicts.
+
+    Args:
+        project_type (str): a string that determines which project (lms or studio) the view is being called in. See the
+            ProjectType enum in plugins/constants.py for valid options
+        slot_namespace (str): a string that specifies the namespace for this slot.
+        slot_name (str): a string that determines which slot the plugin will render content for. These are unique for
+            each project type.
+        raw_context (Dict): the unfiltered context available to the internal view.
+    """
+    aggregate_slot_content = u""
+    slot_functions = _get_cached_slot_functions_for_view(project_type, slot_namespace, slot_name)
+
+    # Each view can pass the approved part of the context, in the context itself.
+    context_allow_list = raw_context.get('context_allow_list', [])
+    if context_allow_list == '*':
+        allowed_context = raw_context
+    else:
+        # The request object is allowed by default, and always available.
+        context_allow_list.extend(COMMON_ALLOW_LIST)
+
+        allowed_context = {
+            key: raw_context.get(key)
+            for key in context_allow_list
+        }
+
+    for (slot_function, plugin_name) in slot_functions:
+        try:
+            # Allow plugins to extend the context for other plugins
+            context = get_plugins_view_context(
+                constants.ProjectType.LMS,
+                plugin_name,
+                raw_context,
+            )
```

Comment:
On some views this will call the plugin's context function twice since they are already called in the view. I understand the goal of "My plugin should be able to add context, and use that context in new HTML/slots without touching edx-platform" though. So that part is fine. 

The `get_plugins_view_context` parameters are `(project_type, view_name, existing_context)` though and this is passing in a `plugin_name` instead of `view_name`. So I don't think this will work. If we're going to do something like this I _would_ like for us to be able to pass in a plugin_name parameter so that we go from
`Allow plugins to extend the context for other plugins`
to
`Allow plugins to extends the context for themselves`

I.e. The only additional context a plugin gets here is bits that they've created. I don't think it makes sense for plugin's to rely on each other. (That's a larger argument though).

Reply:
Passing in the plugin name is intentional, and aims to allow plugins to extend their own context by using their own name, of the context of another plugin. 

If plugina has supports extending the context of pluginb, then in the case that plugina isn't installed, that part of the code just won't run. So, it's not really causing a reliance between plugins. 

That said, I do think this isn't really very useful and it makes sense to just drop it, in favour if just passing the whole context unfiltered context to the plugin. The goal of filtering the context was to pass bare minimum guaranteed context, so that the template context doesn't become something that can break plugins across releases. However, now we have context plugins which are going to break anyway if the context changes, and they rely on a specific context shape.

So we can just pass the entire context to the plugin and then it doesn't need to plug into the platform twice for one view.